### PR TITLE
[Triton/Gluon] [ASM] [HIP] MHA v4: fixes, refactor, new kernel, perf tweaks

### PR DESCRIPTION
**Pull request by jcaraban**
## Motivation

Expand MHA v4 on gfx950 with additional low-precision dense attention recipes, improved quantization and packing paths, and unified recipe dispatch.

## Technical Details

Kernels
- **bugfix**: prevents exp RAW in BF16 kernel epilog
- New BF16 Q/K FP8 P/V kernel (between full bf16 and i8fp8)
- Optimized F8F6, F4F4 and MXFP6, which now cast P to fp6
- All kernels rebased to be `block-sparse` `sorted-dispatch` compatible

Host
- **bugfix**: added zeroed buffer slack for speculative MXFP4 scale reads.
- Added packed V layouts optimized for FP6 probability kernels.
- Split quantization and layout producers into `mha_v4_quant.py`.
- Centralized format, scale-mode, packing, and sparse recipe planning.
- Refactored bench_sage.py around recipe-specific providers and payload accounting.

Minor
- Extended the dispatch manifest with v_pack and refreshed gfx950 code objects.
- Consolidated MXFP8 handling into mha_v4; mha_v4_packed launch options are now keyword-only.
- Updated MHA v4 documentation to include all the above.

## Test Plan

- Run the complete MHA v4 test suite on MI355X/gfx950.
- Validate recipe resolution, stable ABI IDs, scale modes, packing selection, and unsupported combinations.
- Compare quantization and packed layouts against reference implementations.
- Validate BF16/FP8 numerical accuracy, GQA, sparse/dense parity, sparse LUT semantics, ragged tails, and empty rows.
- Validate eager versus torch.compile(fullgraph=True) execution.

## Test Result

- 221 passed in test_mha_v4.py on MI355X/gfx950.
- black==26.3.0 --check aiter/ op_tests/: passed.
- ruff==0.15.7 on all PR-touched Python files: passed.

image

Previous PRs:  https://github.com/ROCm/aiter/pull/4627,  https://github.com/ROCm/aiter/pull/4967,  https://github.com/ROCm/aiter/pull/5005

### CODE REVIEW

File: aiter/ops/triton/quant/sage_attention_quant_wrappers.py [hunk 1]
```diff
@@ -360,6 +360,35 @@ def _pack_v_mxfp4_colmajor_raw_fake(value):
...
+    """Pack MXFP4 V in the token order required by an FP6 P operand.
+
+    The returned payload and E8M0 scale buffers have the same shapes as the
+    canonical col-major packer. Tail source tokens are clamped to the final
+    logical token before quantization.
+    """
+    _, sequence, _, _ = value.shape
+    token = torch.arange(sequence, device=value.device)
+    within_block = token % 64
+    paired = (
```
Comment (valarLip) [inline aiter/ops/triton/quant/sage_attention_quant_wrappers.py]: pure torch impl here?

File: op_tests/test_mha_v4.py
Comment (valarLip) [inline op_tests/test_mha_v4.py:1]: maybe rewrite with https://github.com/ROCm/aiter/tree/main/.claude/skills/aiter-op-test

**Review by valarLip — CHANGES_REQUESTED** [review]

File: aiter/ops/triton/quant/sage_attention_quant_wrappers.py (hunk 1 above)
Comment (jcaraban) [reply aiter/ops/triton/quant/sage_attention_quant_wrappers.py]: Sorry this is dead code 💀  I thought I deleted it, maybe it returned when I merged with https://github.com/ROCm/aiter/pull/5312. Gone now 🧹

File: op_tests/test_mha_v4.py
Comment (jcaraban) [reply op_tests/test_mha_v4.py:1]: pushed a commit that aligns with the skill, although it didn't cause many changes

File: aiter/ops/mha_v4.py [hunk 2]
```diff
@@ -1,109 +1,83 @@
...
-    """Apply hd128 Walsh-Hadamard rotation and pack K in the MXFP4 ASM tile order."""
+__all__ = (
+    "MHA_V4_LOG2E",
+    "AttentionFormat",
+    "AttentionPack",
+    "AttentionScaleMode",
+    "mha_v4",
+    "mha_v4_kv_tile",
+    "mha_v4_packed",
+    "mha_v4_q_multiplier",
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply aiter/ops/mha_v4.py:59]: There were no users yet, but added `mha_v4_mxfp8` back as deprecated in [76b579b](https://github.com/ROCm/aiter/pull/5335/commits/76b579bd53a81992ba46e3aeda523ff1cd70b5fe)

File: csrc/kernels/mha_v4_quant.cu [hunk 3]
```diff
@@ -517,6 +521,205 @@ __global__ void hadamard_rotate_activation_mxfp4_quant_kernel(
...
+        head_tile * kMxfp6VTileBytes + bn * 1536 + lane_group * 24;
+    *reinterpret_cast<packed_t*>(out + data_offset) = packed;
+
+    const int32_t scale_lane = physical_d % 32 + 32 * (kv_block % 2);
+    const int64_t scale_offset =
+        head_tile * kMxfp6VScaleBytes + (kv_block / 2) * 256 + scale_lane * 4 + n;
+    scale[scale_offset] = scale_exp;
+
+    if(group_idx < kMxfp6BufferSlack)
+        out[groups * 24 + group_idx] = 0;
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply csrc/kernels/mha_v4_quant.cu:724]: addressed in [c50e3ba](https://github.com/ROCm/aiter/pull/5335/commits/c50e3ba6ed89dfec665c9019127e64ec9347c512)

File: op_tests/test_mha_v4.py [hunk 4]
```diff
@@ -959,22 +1266,53 @@ def test_mha_v4_raw_compile_parity(q_format, v_format):
 
     assert eager.data_ptr() == eager_out.data_ptr()
     assert compiled.data_ptr() == compiled_out.data_ptr()
-    assert torch.equal(eager, compiled)
+    if q_format == AttentionFormat.BF16 and v_format == AttentionFormat.FP8:
+        torch.testing.assert_close(eager, compiled, atol=0.02, rtol=0.02)
+    else:
+        assert torch.equal(eager, compiled)
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply op_tests/test_mha_v4.py]: addressed in [2ef3e83](https://github.com/ROCm/aiter/pull/5335/commits/2ef3e83fa64abc551850d84cf423e126a58c5b9f)

File: op_tests/test_mha_v4.py [hunk 5]
```diff
@@ -1165,50 +1503,190 @@ def test_mha_v4_sparse_work_table_leaves_uniform_counts_in_raster_order(
...
+    _assert_sparse_matches_dense(sparse, dense)
+
+
+def _assert_sparse_matches_dense(sparse, dense, message=None):
+    """Compare code objects that use different softmax reduction schedules."""
+    cosine = torch.nn.functional.cosine_similarity(
+        sparse.float().flatten(), dense.float().flatten(), dim=0
+    )
+    assert cosine > 0.99, message
     assert torch.isfinite(sparse).all()
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply op_tests/test_mha_v4.py:1807]: will defer to a follow-up PR where I go over all the sparse kernels, this PR is focused on the dense kernels

File: csrc/py_itfs_cu/asm_mha_v4_fwd.cu [hunk 6]
```diff
@@ -634,16 +666,22 @@ PackedMhaV4Shapes validate_packed_mha_v4(const at::Tensor& q,
         TORCH_CHECK(q_descale.numel() == 1 && k_descale.numel() == 1,
                     "INT8/FP8 Q/K descales must be scalar tensors");
```
[the remark this answers is not in the document]
Comment (jcaraban) [reply csrc/py_itfs_cu/asm_mha_v4_fwd.cu:742]: addressed in [86b9591](https://github.com/ROCm/aiter/pull/5335/commits/86b95918e775eb014fc98740948b0d268a5bea97)

**Comment by jcaraban** [comment]
copilot review addressed @Boss2002n , do you have further comments?

_edit:_ I solved other potential issues in the same line of what copilot pointed out

**Review by valarLip — APPROVED** [review]
LGTM

**Comment by Boss2002n** [comment]
> copilot review addressed @Boss2002n , do you have further comments?
> 
> _edit:_ I solved other potential issues in the same line of what copilot pointed out

if we are adding a new kernel - id say pls open a new PR, makes it so much more easier to review it - right now its just a very dense PR and very hard to follow 

atleast for the triton changes

**Review by Boss2002n — CHANGES_REQUESTED** [review]
If possible pls move the triton changes to another PR

**Comment by valarLip** [comment]
> If possible pls move the triton changes to another PR

you are so rude that you don't even look the code

**Comment by jcaraban** [comment]
> if we are adding a new kernel - id say pls open a new PR, makes it so much more easier to review it - right now its just a very dense PR and very hard to follow
> 
> atleast for the triton changes

Sorry @Boss2002n, I agree the PR is dense. There are many kernels under this MHA v4 group and typically one PR needs to address them all together. I also need to pack bugfixes along, as we don't want segfaults in production.

I'm actively migrating from the Triton logic to self-contained HIP. The history behind is that originally we had Triton Sage v1 (i8fp8) and Save v3? (mxfp4) and the ASM kernels have been replacing them. The ASM variants still used the quantization / packing / rotation logic from Triton, but I'm cutting that dependency for better isolated HIP kernels.

Right now this PR only touches [aiter/ops/triton/attention/utils.py‎](https://github.com/ROCm/aiter/pull/5335/changes#diff-0669f716add1b1ae8a2424d70aa525a301f3a3be2d1cfe3588f2c5cf6cede086) with a TODO (for the next PR) and [op_tests/op_benchmarks/triton/bench_sage.py‎](https://github.com/ROCm/aiter/pull/5335/changes#diff-c2388422eb3ea4ef9e8f7fe93daa29eb703cffbc7ef7809deb7fcfa9db4077df) which tests the perf of both Triton and ASM sage kernels so unfortunately always shows large diffs

**Comment by Boss2002n** [comment]
> > if we are adding a new kernel - id say pls open a new PR, makes it so much more easier to review it - right now its just a very dense PR and very hard to follow
> > atleast for the triton changes
> 
> Sorry @Boss2002n, I agree the PR is dense. There are many kernels under this MHA v4 group and typically one PR needs to address them all together. I also need to pack bugfixes along, as we don't want segfaults in production.
> 
> I'm actively migrating from the Triton logic to self-contained HIP. The history behind is that originally we had Triton Sage v1 (i8fp8) and Save v3? (mxfp4) and the ASM kernels have been replacing them. The ASM variants still used the quantization / packing / rotation logic from Triton, but I'm cutting that dependency for better isolated HIP kernels.
> 
> Right now this PR only touches [aiter/ops/triton/attention/utils.py‎](https://github.com/ROCm/aiter/pull/5335/changes#diff-0669f716add1b1ae8a2424d70aa525a301f3a3be2d1cfe3588f2c5cf6cede086) with a TODO (for the next PR) and [op_tests/op_benchmarks/triton/bench_sage.py‎](https://github.com/ROCm/aiter/pull/5335/changes#diff-c2388422eb3ea4ef9e8f7fe93daa29eb703cffbc7ef7809deb7fcfa9db4077df) which tests the perf of both Triton and ASM sage kernels so unfortunately always shows large diffs

Oh no, That makes perfect sense :) 
I assumed the new kernel part after reading the PR title (my apologies)
thanks

**Review by Boss2002n — APPROVED** [review]
